### ISSUE TITLE
Double function call in Numba

### ISSUE THREAD

**Issue by PokhodenkoSA** [IntelPython/numba-dppy#382]
Double function call, wrapper or wrong debug info generation. Blocker for back trace and local variables.

**Comment by akharche** [comment]
`info locals` doesn't show the values of variables when gdb hits breakpoint the first time. Testing the example below I see the strange behavior of debugging with gdb. 

```python
import numpy as np
import numba

@numba.njit(debug=True)
def foo(a):
    b = a + 1
    c = a * 2.34
    d = (a, b, c)
    print(a, b, c, d)
    return b, c, d

r = foo(123)
print(r)
```
When gdb hits breakpoint the first time, the value of **a** function argument is equal to 0. But if I call next (`n` command) several times, gdb hits the breakpoint again and at this time I see that the value of **a** is equal to 123. Also, if I step all the function again I see all the values of local variables.

```bash
$ gdb -q --args python numba_ex.py
Reading symbols from python...
(gdb) break numba_ex.py:6
No source file named numba_ex.py.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (numba_ex.py:6) pending.
(gdb) run
Thread 1 "python" hit Breakpoint 1, __main__::foo$241 () at numba_ex.py:6
6           b = a + 1
(gdb) info locals
d = "\000.\037\302\314m\017| \375\336\342\377\177\000\000@\350\000\334\377\177\000"
a = 0
b = <optimized out>
c = <optimized out>
(gdb)
```

```bash
(gdb) n
7           c = a * 2.34
(gdb) n
8           d = (a, b, c)
(gdb) n
9           print(a, b, c, d)
(gdb) n
10          return b, c, d
(gdb) n
6           b = a + 1
(gdb) info locals
d = '\000' <repeats 23 times>
a = 123
b = 0
c = 0

(gdb) n
7           c = a * 2.34
(gdb) n
8           d = (a, b, c)
(gdb) n
9           print(a, b, c, d)
(gdb) info locals
d = "{\000\000\000\000\000\000\000|\000\000\000\000\000\000\000\205\353Q\270\036\375q@"
a = 123
b = 124
c = 287.81999999999999

```

**Comment by akharche** [comment]
Fix from Numba: https://github.com/numba/numba/pull/7116

**Comment by stuartarchibald** [comment]
I think https://github.com/numba/numba/pull/7177/commits/2d637845ebe305c8b12d451a55c71b8c1ba377fd from https://github.com/numba/numba/pull/7177 is going to be needed as well for dealing with refcounted types.

**Comment by akharche** [comment]
Tested [numba/numba#7177](https://github.com/numba/numba/pull/7177) 
```python
import numba
import numpy as np

@numba.njit(debug=True)
def foo(a, arr):
    b = arr[3] + a
    c = b * 2.34
    d = 10 * c
    return d, arr


arr = np.arange(10)
r = foo(123, arr)
print(r, a)

```

```bash
(numba54) akharche@ansatnuc05:/localdisk/work/akharche$ gdb-oneapi -q --args python numba_ex.py
Reading symbols from python...
(gdb) break numba_ex.py:6
No source file named numba_ex.py.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (numba_ex.py:6) pending.
(gdb) run
Starting program: /localdisk/work/akharche/miniconda3/envs/numba54/bin/python numba_ex.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[Detaching after fork from child process 3625045]
[Detaching after fork from child process 3625046]
[Detaching after fork from child process 3625047]

Breakpoint 1, __main__::foo$241 () at numba_ex.py:6
6           b = arr[3] + a
(gdb) info locals
a = 0
arr = {meminfo = 0x0, parent = 0x0, nitems = 0, itemsize = 0, data = 0x0, shape = {0}, strides = {0}}
b = 0
c = 0
d = 0

(gdb) n
7           c = b * 2.34
(gdb) n
8           d = 10 * c
(gdb) info locals
a = 0
arr = {meminfo = 0x5555564753a0 "\002", parent = 0x7fffdf36fb70 "\005", nitems = 10, itemsize = 8, data = 0x555556203d90, shape = {10}, strides = {8}}
b = 0
c = 294.83999999999997
d = 0
(gdb) whatis arr
type = array(int64, 1d, C) ({i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64]})

```

**Comment by PokhodenkoSA** [comment]
Fixed in Numba 0.55.

**Comment by stuartarchibald** [comment]
> Fixed in Numba 0.55.

Thanks for confirming.
